query-engine/connector-test-kit-rs/query-engine-tests/tests/writes/top_level_mutations/create_many.rs

```diff
@@ -294,6 +295,181 @@ mod create_many {
+                  { req: 2, opt_default_static: 2 },
+                  { req: 3, req_default_static: 3 },
+                  { req: 4, opt: 8 },
+                  { req: 5, opt: null },
+                ]
+              ) {
+                count
+              }
+            }"#
+        );
+
+        let json = runner.get_metrics().to_json(Default::default());
+        let counter = metrics::get_counter(&json, PRISMA_DATASOURCE_QUERIES_TOTAL);
+
+        match runner.max_bind_values() {
+            Some(x) if x >= 18 => assert_eq!(counter, 3), // 1 createMany queries (BEGIN/COMMIT are counted)
+            // Some queries are being split because of `QUERY_BATCH_SIZE` being set to `10` in dev.
+            Some(_) => assert_eq!(counter, 4), // 2 createMany queries (BEGIN/COMMIT are counted)
+            _ => panic!("Expected max bind values to be set"),
+        }
+
+        Ok(())
+    }
+
+    // LibSQL & co are ignored because they don't support metrics
+    #[connector_test(schema(schema_7), only(Sqlite("3")))]
+    async fn create_many_by_shape_counter_3(runner: Runner) -> TestResult<()> {
+        use query_engine_metrics::PRISMA_DATASOURCE_QUERIES_TOTAL;
+
+        // Generated queries:
+        // INSERT INTO `main`.`Test` ( `req_default_static`, `req`, `opt_default`, `opt_default_static` ) VALUES (?, ?, ?, ?) params=[1,6,3,1]
+        // INSERT INTO `main`.`Test` ( `opt`, `req`, `req_default_static`, `opt_default_static` ) VALUES (null, ?, ?, ?), (null, ?, ?, ?), (null, ?, ?, ?) params=[1,1,1,2,1,2,3,3,1]
+        // INSERT INTO `main`.`Test` ( `opt`, `req`, `req_default_static`, `opt_default_static` ) VALUES (?, ?, ?, ?), (?, ?, ?, ?) params=[8,4,1,1,null,5,1,1]
+        // Note: The first two queries are split because QUERY_BATCH_SIZE is set to 10. In production, only two queries would be generated for this example.
+        run_query!(
+            &runner,
+            r#"mutation {
+              createManyTest(
+                data: [
+                  { req: 1 }
+                  { req: 2, opt_default_static: 2 },
+                  { req: 3, req_default_static: 3 },
+                  { req: 4, opt: 8 },
+                  { req: 5, opt: null },
+                  { req: 6, opt_default: 3 },
+                ]
+              ) {
+                count
+              }
+            }"#
+        );
+
+        let json = runner.get_metrics().to_json(Default::default());
+        let counter = metrics::get_counter(&json, PRISMA_DATASOURCE_QUERIES_TOTAL);
+
+        match runner.max_bind_values() {
+            Some(x) if x > 21 => assert_eq!(counter, 4), // 3 createMany queries in total (BEGIN/COMMIT are counted)
+            // Some queries are being split because of `QUERY_BATCH_SIZE` being set to `10` in dev.
+            Some(_) => assert_eq!(counter, 5), // 3 createMany queries in total (BEGIN/COMMIT are counted)
+            _ => panic!("Expected max bind values to be set"),
+        }
+
+        Ok(())
+    }
```

Comment:
Okay, so, I think this `Some(_)` line makes sense in that the max_bind_value isn't affecting what's happening so we get the actual 5 composed of 3 `createMany`, 1 `BEGIN`, and 1 `COMMIT`.

1. Where is this `21` coming from for the above branch?
I see we have the following `max_bind_values` default for sqlite, so 21 also represents `QUERY_BATCH_SIZE`
	```rs
	SqlFamily::Sqlite => 999
	```

	So then with the higher batch size we can merge the above two queries
	```
	INSERT INTO `main`.`Test` (
		`req_default_static`, `req`, `opt_default`, `opt_default_static`
	) VALUES (?, ?, ?, ?) params=[1,6,3,1]
	INSERT INTO `main`.`Test` (
	`opt`, `req`, `req_default_static`, `opt_default_static`
	) VALUES (null, ?, ?, ?), (null, ?, ?, ?), (null, ?, ?, ?) params=[1,1,1,2,1,2,3,3,1]
	```
	
	but still also curious as to _why_ `21`

Reply:
21 is the number of query params for the generated `INSERT` statements. If you have a `max_bind_value` that's lower than the number of params the `createMany` operation generates, then it will not split the `createMany` operation into more.

<img width="1380" alt="image" src="https://github.com/prisma/prisma-engines/assets/25233379/2f209490-02f6-45d8-9fc9-4ac455099963">

Reply:
Cool, thanks for explaining!